client/state/query-client-ssr.ts

```diff
@@ -1,13 +1,44 @@
-import { QueryClient, QueryCache } from 'react-query';
+import { QueryClient, QueryCache, dehydrate, QueryOptions, QueryKey } from 'react-query';
 import { MAX_AGE, BASE_STALE_TIME } from 'calypso/state/initial-state';
+import type { Query, QueryState } from 'react-query/types/core/query';
 
-const sharedCache = new QueryCache();
+class QueryClientSSR extends QueryClient {
+	fetchedQueryKeys = new Set< string >();
+}
+
+class QueryCacheSSR extends QueryCache {
+	build< TQueryFnData, TError, TData, TQueryKey extends QueryKey >(
+		client: QueryClient | QueryClientSSR,
+		options: QueryOptions< TQueryFnData, TError, TData, TQueryKey >,
+		state?: QueryState< TData, TError >
+	): Query< TQueryFnData, TError, TData, TQueryKey > {
+		const query = super.build( client, options, state );
+		if ( client instanceof QueryClientSSR && ( options as any ).enabled !== false ) {
+			client.fetchedQueryKeys.add( query.queryHash );
+		}
+		return query;
+	}
+}
+
+const sharedCache = new QueryCacheSSR();
 
 export function createQueryClientSSR() {
-	const queryClient = new QueryClient( {
+	const queryClient = new QueryClientSSR( {
 		defaultOptions: { queries: { cacheTime: MAX_AGE, staleTime: BASE_STALE_TIME } },
 		queryCache: sharedCache,
 	} );
 
 	return queryClient;
 }
+
+export function dehydrateQueryClient( queryClient: QueryClientSSR ) {
+	if ( ! queryClient ) {
+		return null;
+	}
+
+	return dehydrate( queryClient, {
+		shouldDehydrateQuery: ( { queryHash: queryKey } ) => {
+			return queryClient.fetchedQueryKeys.has( queryKey );
```

Comment:
There's some naming confusion about `queryHash` and `queryKey`. Both fields exist on a query object, and are different from each other. Let's use consistent naming. The `queryHash` shouldn't be assigned to a variable named `queryKey`, and the `fetchedQueryKeys` set should be `fetchedQueryHashes`, because it contains `queryHash`es.